shared/src/main/java/de/pnp/manager/network/state/StateMachine.java

```diff
@@ -0,0 +1,73 @@
+package de.pnp.manager.network.state;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+public class StateMachine<Event> {
+
+    protected Set<State> states;
+    protected State currentState;
+    protected Collection<Transition<Event>> transitions;
+    protected Consumer<Event> onNoTransition;
+
+    public StateMachine(Set<State> states, State start) {
+        this.states = states;
+        this.currentState = start;
+        this.transitions = new ArrayList<>();
+        this.onNoTransition = null;
+    }
+
+    public boolean fire(Event event) {
+
+        for (Transition<Event> transition : transitions) {
+            if (
+                    transition.from == currentState &&
+                            transition.eventCheck.apply(event) &&
+                            states.contains(transition.to)
+            ) {
+                if (transition.eventHandler != null) {
+                    transition.eventHandler.accept(event);
+                }
+                this.currentState = transition.to;
+                return true;
+            }
+        }
+
+        if (onNoTransition != null) {
+            onNoTransition.accept(event);
+        }
+
+        return false;
+    }
+
+    public void registerTransition(State from, State to, @NotNull Function<Event, Boolean> eventCheck, Consumer<Event> eventHandler) {
```

Comment:
should we check that we do not have two transitions with the same fromState and event?

Reply:
With this implementation, we can say if two transition react to the same event.
We use a function to check if a transition reacts to the given event.
I would say that it is no problem.
It would always take the transition which was first registered.
In this way, it it determenistic.
You could use this feature to have a specific case for an event and one default case.